### ISSUE TITLE
Prometheus and Alertmanager objects cannot have the same name

### ISSUE THREAD

**Issue by snarlysodboxer** [coreos/prometheus-operator#93]
I was thinking it'd make sense to have a `Prometheus`, `Alertmanager`, (not `AlertManager`?) and `ServiceMonitor` object definition each named `frontend`, the three being related, yet already differentiated by their kind. However, since the Operator looks for ConfigMaps named `<(prometheus name|alertmanager name)>`, I assume it would break.

What about an option to specify the `configMap`'s name/label?

**Comment by fabxc** [comment]
In our design so far we assumed naming to be like `prometheus-frontend`. I can see how one wants to skip the prefix though as it's given by the the resource kind.

One option would be specifying configmaps explicitly as you described. Another or additional way may be generally apply the prefix on the operator side for created objects such as configmaps or stateful sets (@brancz). That might cause confusion due to its implicitness though.

A general note, you probably don't want to have one Alertmanager for every Prometheus server.
It's a highly available component that's meant to have a global view of all alerts generated by different Prometheus setups.
Typically, you need/want only one Alertmanager setup for your entire organisation.

**Comment by snarlysodboxer** [comment]
@fabxc Thanks for that tip about my Alertmanager setup, that make a lot of sense, I'll adjust my setup!

While I still think it's a good idea to be able to have the TPR objects named the same, I did discover that a simple name of `frontend` for the Prometheus object causes more ambiguity in the pods names than I prefer. E.G. I end up with a list of pods like:
```shell
alertmanager-0
alertmanager-1
frontend-0
frontend-1
grafana-2200421-63c1m
heapster-1694355377-rlvlw
influxdb-1824287510-l7g7x
prometheus-operator-877765330-5gng4
```
and the name `frontend-0` doesn't tell me what the pod does. I've switched back to `frontend-prometheus` for now.

Maybe your idea of applying a prefix is a good idea. Or maybe a postfix? E.G.:
```
frontend-prometheus
backend-prometheus
```
vs.
```
prometheus-frontend
prometheus-backend
```
The latter make is sound like a single prometheus setup with frontend and backend components.

The implicitness would be ok if documented IMO, especially since you won't be expecting to use pod names in any k8s resource config I can think of.

**Comment by brancz** [comment]
I like this idea of

```
frontend-prometheus
backend-prometheus
```

It is somewhat similar to a `Deployment` adding a suffix to a `ReplicaSet`.

**Comment by fabxc** [comment]
Mh, I'm a prefix person if it comes to these things to be honest. But that might just be personal taste.

I'm a bit wary on implicitness. It would certainly be a breaking change, right? (Which is okay in general).
There's probably an unbreaking upgrade path in some way, but coupling implicitness with implicit handling of two variants won't end well.

**Comment by brancz** [comment]
This is going to be breaking yes, which is why we should jump to a `v0.3.0` for this. I'm fine with going with a prefix. We should go through all namings that we may potentially break ever as that is one thing that can give us huge headaches, as we have no way of migrating `Pod`s from an old `StatefulSet` to a new one if the names don't match. The only way to migrate then is to delete the `PersistentVolume`s but keep the storage in tact, and then have the new `PersisentVolume`s pick up the correct underlying storage again. I'm not sure this is possible with storage provisioning.

Just something to keep in mind, a breaking path is ok for now, but we need to carefully consider paths that could give us headaches like the above mentioned.

**Comment by brancz** [comment]
We have now implemented this and went the prefix way. Closed by #108
